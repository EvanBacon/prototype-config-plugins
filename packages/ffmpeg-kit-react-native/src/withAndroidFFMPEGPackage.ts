import {
  createGeneratedHeaderComment,
  MergeResults,
  removeGeneratedContents,
} from "@expo/config-plugins/build/utils/generateCode";
import { ExpoConfig } from "expo/config";
import {
  ConfigPlugin,
  withGradleProperties,
  withProjectBuildGradle,
} from "expo/config-plugins";

export const withAndroidFFMPEGPackage: ConfigPlugin<string | undefined> = (
  config,
  packageName,
) => {
  config = withProjectBuildGradle(config, (config) => {
    if (config.modResults.language === "groovy") {
      config.modResults.contents = addPackageName(
        config.modResults.contents,
        packageName,
      );
    } else {
      throw new Error(
        "Cannot add camera maven gradle because the build.gradle is not groovy",
      );
    }
    return config;
  });

  config = withLibCppSharedSo(config);

  return config;
};

export function addPackageName(src: string, packageName?: string): string {
  const tag = "ffmpeg-kit-react-native-package";
  const gradleMaven = packageName
    ? `ext { ffmpegKitPackage = "${packageName}" }`
    : "";
  return appendContents({
    tag,
    src,
    newSrc: gradleMaven,
    comment: "//",
  }).contents;
}

// Fork of config-plugins mergeContents, but appends the contents to the end of the file.
function appendContents({
  src,
  newSrc,
  tag,
  comment,
}: {
  src: string;
  newSrc: string;
  tag: string;
  comment: string;
}): MergeResults {
  const header = createGeneratedHeaderComment(newSrc, tag, comment);
  if (!src.includes(header)) {
    // Ensure the old generated contents are removed.
    let sanitizedTarget = removeGeneratedContents(src, tag);
    if (sanitizedTarget) sanitizedTarget += "\n";

    const contentsToAdd = [
      // @something
      header,
      // contents
      newSrc,
      // @end
      `${comment} @generated end ${tag}`,
    ].join("\n");

    return {
      contents: (sanitizedTarget ?? src) + contentsToAdd,
      didMerge: true,
      didClear: !!sanitizedTarget,
    };
  }
  return { contents: src, didClear: false, didMerge: false };
}

// This is required if you have several libraries which include libc++_shared.so as a dependency.
// See https://github.com/tanersener/ffmpeg-kit/wiki/Tips#2-depending-another-android-library-containing-libc_sharedso
function withLibCppSharedSo(config: ExpoConfig): ExpoConfig {
  return withGradleProperties(config, (config) => {
    config.modResults.push({
      type: "property",
      key: "android.packagingOptions.pickFirsts",
      value:
        "lib/x86/libc++_shared.so,lib/x86_64/libc++_shared.so,lib/armeabi-v7a/libc++_shared.so,lib/arm64-v8a/libc++_shared.so",
    });
    return config;
  });
}
