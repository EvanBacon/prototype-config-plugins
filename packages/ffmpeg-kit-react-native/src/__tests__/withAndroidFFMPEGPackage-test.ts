import { getFixture } from "../../../../fixtures/getFixtures";
import { addPackageName } from "../withAndroidFFMPEGPackage";

describe(addPackageName, () => {
  it(`adds undefined package`, () => {
    expect(
      addPackageName(getFixture("build.gradle"), undefined),
    ).toMatchSnapshot();
  });
  it(`redefines the package`, () => {
    let value = addPackageName(getFixture("build.gradle"), "video");
    value = addPackageName(value, "audio");
    value = addPackageName(value, undefined);
    value = addPackageName(value, "min");
    expect(value).toMatchSnapshot();
  });
});
