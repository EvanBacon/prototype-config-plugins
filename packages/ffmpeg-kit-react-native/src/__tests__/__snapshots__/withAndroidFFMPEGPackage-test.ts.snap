// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPackageName adds undefined package 1`] = `
"
buildscript {
    ext {
        buildToolsVersion = "29.0.2"
        minSdkVersion = 21
        compileSdkVersion = 29
        targetSdkVersion = 29
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath("com.android.tools.build:gradle:3.5.3")
    }
}

allprojects {
    repositories {
        mavenLocal()
        maven {
            url("$rootDir/../node_modules/react-native/android")
        }
        maven {
            url("$rootDir/../node_modules/jsc-android/dist")
        }
        google()
        jcenter()
        maven { url 'https://www.jitpack.io' }
    }
}
// @generated begin ffmpeg-kit-react-native-package - expo prebuild (DO NOT MODIFY) sync-da39a3ee5e6b4b0d3255bfef95601890afd80709

// @generated end ffmpeg-kit-react-native-package"
`;

exports[`addPackageName redefines the package 1`] = `
"
buildscript {
    ext {
        buildToolsVersion = "29.0.2"
        minSdkVersion = 21
        compileSdkVersion = 29
        targetSdkVersion = 29
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath("com.android.tools.build:gradle:3.5.3")
    }
}

allprojects {
    repositories {
        mavenLocal()
        maven {
            url("$rootDir/../node_modules/react-native/android")
        }
        maven {
            url("$rootDir/../node_modules/jsc-android/dist")
        }
        google()
        jcenter()
        maven { url 'https://www.jitpack.io' }
    }
}
// @generated begin ffmpeg-kit-react-native-package - expo prebuild (DO NOT MODIFY) sync-3d2afee3bf3e62c8e7e2f5faaff1396dc491d7de
ext { ffmpegKitPackage = "min" }
// @generated end ffmpeg-kit-react-native-package"
`;
